app/src/main/java/com/example/unsplashphotos/ui/GalleryFragment.kt

```diff
@@ -0,0 +1,76 @@
+package com.example.unsplashphotos.ui
+
+import android.annotation.SuppressLint
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.viewModels
+import androidx.lifecycle.lifecycleScope
+import androidx.navigation.fragment.findNavController
+import androidx.recyclerview.widget.GridLayoutManager
+import androidx.recyclerview.widget.RecyclerView
+import com.example.unsplashphotos.data.model.Photo
+import com.example.unsplashphotos.databinding.FragmentGalleryBinding
+import com.example.unsplashphotos.ui.adapter.PhotoAdapter
+import dagger.hilt.android.AndroidEntryPoint
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+@AndroidEntryPoint
+class GalleryFragment : Fragment() {
+
+    private lateinit var photoAdapter: PhotoAdapter
+    private lateinit var binding: FragmentGalleryBinding
+    private val photoViewModel by viewModels<PhotoViewModel>()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        photoAdapter = PhotoAdapter{ selectedPhoto: Photo, extra ->
+            val action = GalleryFragmentDirections.actionGalleryFragmentToPhotoFullScreenFragment(selectedPhoto.id)
+            findNavController().navigate(action)
+        }
+
+        photoAdapter.stateRestorationPolicy =
+            RecyclerView.Adapter.StateRestorationPolicy.PREVENT_WHEN_EMPTY
+
+    }
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        // Inflate the layout for this fragment
+        binding = FragmentGalleryBinding.inflate(inflater)
+        return binding.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        setupPhotoRecyclerView()
+    }
+
+    private fun setupPhotoRecyclerView() {
+        binding.photosRecyclerView.adapter = photoAdapter
+        photoViewModel
+        val spanCount = 1
+        binding.photosRecyclerView.layoutManager =
+            GridLayoutManager(requireContext(), spanCount, GridLayoutManager.VERTICAL, false)
+
+        displayPhotos()
+    }
+
+    @SuppressLint("NotifyDataSetChanged")
+    private fun displayPhotos() {
+        lifecycleScope.launch {
+            withContext(Dispatchers.IO) {
+                photoViewModel.fetchPhotos().collectLatest {
```

Comment:
launch this from Viewmodel scope and collect it safely from Fragment scope

Reply:
https://medium.com/androiddevelopers/a-safer-way-to-collect-flows-from-android-uis-23080b1f8bda

Reply:
https://developer.android.com/topic/libraries/architecture/coroutines